extensions/azure/processors/FetchAzureDataLakeStorage.cpp

```diff
@@ -0,0 +1,122 @@
+      ->supportsExpressionLanguage(true)
+      ->build());
+
+const core::Property FetchAzureDataLakeStorage::NumberOfRetries(
+    core::PropertyBuilder::createProperty("Number of Retries")
+      ->withDescription("The number of automatic retries to perform if the download fails.")
+      ->withDefaultValue<uint64_t>(0)
+      ->supportsExpressionLanguage(true)
+      ->build());
+
+const core::Relationship FetchAzureDataLakeStorage::Success("success", "Files that have been successfully fetched from Azure storage are transferred to this relationship");
+const core::Relationship FetchAzureDataLakeStorage::Failure("failure", "In case of fetch failure flowfiles are transferred to this relationship");
+
+void FetchAzureDataLakeStorage::initialize() {
+  // Add new supported properties
+  setSupportedProperties({
+    AzureStorageCredentialsService,
+    FilesystemName,
+    DirectoryName,
+    FileName,
+    RangeStart,
+    RangeLength,
+    NumberOfRetries
+  });
+  // Set the supported relationships
+  setSupportedRelationships({
+    Success,
+    Failure
+  });
+}
+
+std::optional<storage::FetchAzureDataLakeStorageParameters> FetchAzureDataLakeStorage::buildFetchParameters(
+    const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::FlowFile>& flow_file) {
+  storage::FetchAzureDataLakeStorageParameters params;
+  if (!setCommonParameters(params, *context, flow_file)) {
+    return std::nullopt;
+  }
+
+  std::string value;
+  if (context->getProperty(RangeStart, value, flow_file)) {
+    params.range_start = std::stoull(value);
+    logger_->log_debug("Range Start property set to %llu", *params.range_start);
+  }
+
+  if (context->getProperty(RangeLength, value, flow_file)) {
+    params.range_length = std::stoull(value);
+    logger_->log_debug("Range Length property set to %llu", *params.range_length);
+  }
+
+  if (context->getProperty(NumberOfRetries, value, flow_file)) {
+    params.number_of_retries = std::stoull(value);
+    logger_->log_debug("Number Of Retries property set to %llu", *params.number_of_retries);
+  }
+
+  return params;
+}
+
+void FetchAzureDataLakeStorage::onTrigger(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) {
+  logger_->log_debug("FetchAzureDataLakeStorage onTrigger");
+  std::shared_ptr<core::FlowFile> flow_file = session->get();
+  if (!flow_file) {
+    context->yield();
+    return;
+  }
+
+  const auto params = buildFetchParameters(context, flow_file);
+  if (!params) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  WriteCallback callback(azure_data_lake_storage_, *params, logger_);
+  session->write(flow_file, &callback);
+
+  if (callback.getResult() == std::nullopt) {
+    logger_->log_error("Failed to fetch file '%s' from Azure Data Lake storage", params->filename);
+    session->transfer(flow_file, Failure);
```

Comment:
when the fetch fails, we return 0 in the `WriteCallback` which sets the flowfile's size to 0, then we send this flowfile to Failure, do we want this, or do we want to forward the original flowfile to failure (and maybe create a new one for the fetched content)?

Reply:
Good point, I think it's better to have the original flowfile forwarded to the failure relationship and as I checked it works that way in NiFi too. I updated the processor to create a new flowfile for the success case and forward the original in failure cases in 172d2a4f893238d7a9818b812cc0b7b671c3cd0b